lib/vu_state.go

```diff
@@ -73,68 +73,39 @@ type State struct {
 	GetScenarioGlobalVUIter func() uint64
 }
 
-// CloneTags makes a copy of the tags map and returns it.
-func (s *State) CloneTags() map[string]string {
-	return s.Tags.Clone()
-}
-
-// TagMap is a safe-concurrent Tags lookup.
-type TagMap struct {
-	m     map[string]string
+// VUStateTags wraps the current VU's tags and ensures a thread-safe way to
+// access and modify them exists. This is necessary because the VU tags and
+// metadata can be modified from the JS scripts via the `vu.tags` API in the
+// `k6/execution` built-in module.
+type VUStateTags struct {
 	mutex sync.RWMutex
+	tags  *metrics.TagSet
+	// TODO: Add metadata map[string]string
 }
 
-// NewTagMap creates a TagMap,
-// if a not-nil map is passed then it will be used as the internal map
-// otherwise a new one will be created.
-func NewTagMap(m map[string]string) *TagMap {
-	if m == nil {
-		m = make(map[string]string)
+// NewVUStateTags initializes a new VUStateTags and returns it. It's important
+// that tags is not nil and initialized via metrics.Registry.RootTagSet().
+func NewVUStateTags(tags *metrics.TagSet) *VUStateTags {
+	if tags == nil {
+		panic("the metrics.TagSet must be initialized for creating a new lib.VUStateTags")
 	}
-	return &TagMap{
-		m:     m,
+	return &VUStateTags{
 		mutex: sync.RWMutex{},
+		tags:  tags,
+		// metadata is intentionally nil by default
 	}
 }
 
-// Set sets a Tag.
-func (tg *TagMap) Set(k, v string) {
-	tg.mutex.Lock()
-	defer tg.mutex.Unlock()
-	tg.m[k] = v
-}
-
-// Get returns the Tag value and true
-// if the provided key has been found.
-func (tg *TagMap) Get(k string) (string, bool) {
-	tg.mutex.RLock()
-	defer tg.mutex.RUnlock()
-	v, ok := tg.m[k]
-	return v, ok
-}
-
-// Len returns the number of the set keys.
-func (tg *TagMap) Len() int {
+// GetCurrentValues returns the value of the VU tags in a thread-safe way.
+func (tg *VUStateTags) GetCurrentValues() *metrics.TagSet {
```

Comment:
Hmm, I am maybe fine with getting rid of `Get`, but I kind of like `Current`, since it implies that the value can change, and so implies that there is some kind of locking... :thinking: But even the `Get` part kind of makes sense in the context of this object that is basically there to ensure thread safety :thinking: It maybe makes a bit more sense in the context of the unindexable tags PR (which I really hope to publish the WIP version today :crossed_fingers:), where I have this `TagsAndMeta` struct and the methods look like this:
```go
// metrics.TagsAndMeta
type TagsAndMeta struct {
	Tags     *TagSet
	Metadata map[string]string // could be nil
}

// GetCurrentValues returns the current value of the VU tags and a copy of the
// metadata (if any) in a thread-safe way.
func (tg *VUStateTags) GetCurrentValues() metrics.TagsAndMeta {
	tg.mutex.RLock()
	defer tg.mutex.RUnlock()
	return tg.tagsAndMeta.Clone()
}

// Modify allows the thread-safe modification of the current VU tags and metadata.
func (tg *VUStateTags) Modify(callback func(tagsAndMeta *metrics.TagsAndMeta)) {
	tg.mutex.Lock()
	defer tg.mutex.Unlock()
	callback(tg.tagsAndMeta)
}
```

In this context, when one of the methods returns a pointer and the other does not, I think even `Get` has value.

Reply:
I dislike the current name because it is long and IMO adds very little info. 

I kind of feel like `Values` will be misinterpreted as well. But my suggestion of `Copy` also seems easy to misinterpret so :shrug: 

I am kind of for just leaving it as is - we are going to come back to this code in the future again either way so  we can discuss the better name then.